From 58acf9d96a07641eafe4bfa4fe19e083fa2dd993 Mon Sep 17 00:00:00 2001
From: Abhishek <apatgaonkar@habana.ai>
Date: Fri, 30 Aug 2024 09:13:58 +0300
Subject: [PATCH] [FSW-1175] fs1: load recovery images

Based on recovery image counts in manifest this patch
loads, verifies and boots recovery images.

Change-Id: Ic5ba8153055b458e1dbca3e9e062a8808e81305c
---

diff --git a/common/boot_desc.c b/common/boot_desc.c
index edea392..dd620bc 100644
--- a/common/boot_desc.c
+++ b/common/boot_desc.c
@@ -221,7 +221,7 @@
 }
 
 uint32_t boot_desc_get_img_desc(enum hl_fw_manifest_type manifest_type, int boot_desc_type,
-				struct boot_desc *desc)
+				struct boot_desc *desc, uint32_t index)
 {
 	uint32_t i;
 	struct hl_fw_img_desc *p_loc_img_desc = NULL;
@@ -243,6 +243,9 @@
 		if (p_loc_img_desc->magic.type != boot_desc_type)
 			continue;
 
+		if (index-- != 0)
+			continue;
+
 		return boot_desc_copy_desc(desc, p_loc_img_desc);
 	}
 
diff --git a/common/boot_desc.h b/common/boot_desc.h
index 79334bc..ef414e3 100644
--- a/common/boot_desc.h
+++ b/common/boot_desc.h
@@ -31,7 +31,7 @@
 
 int boot_desc_init(enum hl_fw_manifest_type manifest_type);
 uint32_t boot_desc_get_img_desc(enum hl_fw_manifest_type manifest_type, int boot_desc_type,
-				struct boot_desc *desc);
+				struct boot_desc *desc, uint32_t index);
 bool boot_rom_boot_desc_init_done(enum hl_fw_manifest_type manifest_type);
 uint32_t boot_rom_verify_img_desc_address_range(struct boot_desc *desc,
 						enum boot_rom_progress img_sts_base,
diff --git a/common/boot_flow.c b/common/boot_flow.c
index 324a5c1..19758a2 100644
--- a/common/boot_flow.c
+++ b/common/boot_flow.c
@@ -34,7 +34,7 @@
 	boot_rom_status_update_bringup(img_sts_base, PROGRESS_BRINGUP_START);
 
 	/* get image descriptor from the manifest */
-	if (boot_desc_get_img_desc(manifest_idx, img_type, desc) != BOOT_ROM_SUCCESS) {
+	if (boot_desc_get_img_desc(manifest_idx, img_type, desc, 0) != BOOT_ROM_SUCCESS) {
 		boot_rom_status_update_bringup(img_sts_base, PROGRESS_BRINGUP_NO_IMG_DESC);
 		return BOOT_ROM_FAIL;
 	}
@@ -107,7 +107,7 @@
 	void (*ppboot)();
 	struct boot_desc desc = {0};
 
-	if (boot_desc_get_img_desc(man_idx, HL_FW_IMG_PPBOOT, &desc) == BOOT_ROM_SUCCESS) {
+	if (boot_desc_get_img_desc(man_idx, HL_FW_IMG_PPBOOT, &desc, 0) == BOOT_ROM_SUCCESS) {
 		/* No need to verify as image validation is done in previous boot */
 		if (desc.start_addr) {
 			/* TODO: Start CPLD watchdog timer */
@@ -134,7 +134,7 @@
 	if (!boot_rom_boot_flow_is_manifest_valid(man_type))
 		return;
 
-	if (boot_desc_get_img_desc(man_idx, HL_FW_IMG_PPBOOT, &desc) != BOOT_ROM_SUCCESS)
+	if (boot_desc_get_img_desc(man_idx, HL_FW_IMG_PPBOOT, &desc, 0) != BOOT_ROM_SUCCESS)
 		return;
 
 	/* TODO: MemCopy to PSOC FW SRAM if required */
@@ -227,3 +227,48 @@
 
 	return boot_desc_get_recovery_img_count(manifest_idx);
 }
+
+uint32_t boot_rom_boot_flow_load_recovery_images(enum boot_rom_manifests recovery_manifest,
+						  uint32_t img_count,
+						  enum boot_rom_progress img_sts_base)
+{
+	enum hl_fw_manifest_type manifest_idx = primary_manifest_idx;
+	struct boot_desc desc = {0};
+	uint32_t i;
+	int err;
+	void (*recovery)();
+
+	if (recovery_manifest == BOOT_ROM_SECONDARY)
+		manifest_idx = (primary_manifest_idx + 1) % HL_FW_SPI_MANIFEST_MAX_NUM;
+
+	/* get image descriptor from the manifest */
+	for (i = 0; i < img_count; i++) {
+		if (boot_desc_get_img_desc(manifest_idx, HL_FW_IMG_RECOVERY, &desc, i)
+			!= BOOT_ROM_SUCCESS) {
+			boot_rom_status_update_bringup(img_sts_base, PROGRESS_BRINGUP_NO_IMG_DESC);
+			return BOOT_ROM_FAIL;
+		}
+
+		/* Validate Image descriptor start address and other ranges */
+		if (boot_rom_verify_img_desc_address_range(&desc, img_sts_base, HL_FW_IMG_RECOVERY)
+		!= BOOT_ROM_SUCCESS)
+			return BOOT_ROM_FAIL;
+
+		err = boot_rom_spi_dma(desc.img_size, desc.spi_offset, desc.loadaddr,
+				false, false, 0);
+		if (boot_rom_process_dma_err(err, img_sts_base))
+			return BOOT_ROM_FAIL;
+
+		/* TODO: Validate recovery image hash */
+
+		if (desc.start_addr) {
+			/* TODO: Start CPLD watchdog timer */
+			recovery = (void *)((uint32_t)boot_rom_get_arc_addr(desc.start_addr));
+			boot_rom_status_update(PROGRESS_BOOT_RECOVERY_IMG_IMG0_START);
+			recovery();
+		}
+	}
+
+	boot_rom_status_update(PROGRESS_BOOT_RECOVERY_IMG_IMG0_FINISH);
+	return BOOT_ROM_SUCCESS;
+}
diff --git a/common/boot_flow.h b/common/boot_flow.h
index 2efabad..28b7bca 100644
--- a/common/boot_flow.h
+++ b/common/boot_flow.h
@@ -6,6 +6,8 @@
 #ifndef _BOOT_FLOW_H_
 #define _BOOT_FLOW_H_
 
+#include <progress_bits.h>
+
 uint32_t boot_rom_boot_flow_prefetch_img(enum hl_fw_img_type img_type);
 void boot_rom_boot_flow_boot_ppboot(enum hl_fw_manifest_type man_idx);
 void boot_rom_boot_flow_boot_ppboot_sub(enum boot_rom_manifests man_type);
@@ -15,5 +17,8 @@
 bool boot_rom_boot_flow_is_secondary_manifest_valid(void);
 void boot_rom_boot_flow_wait_for_dma_finish(void);
 uint32_t boot_rom_boot_flow_get_recovery_img_count(enum boot_rom_manifests recovery_manifest);
+uint32_t boot_rom_boot_flow_load_recovery_images(enum boot_rom_manifests recovery_manifest,
+						  uint32_t img_count,
+						  enum boot_rom_progress img_sts_base);
 
 #endif /* _BOOT_FLOW_H_ */
diff --git a/common/recovery.c b/common/recovery.c
index 8557032..be9baf8 100644
--- a/common/recovery.c
+++ b/common/recovery.c
@@ -42,7 +42,11 @@
 		return BOOT_ROM_IMAGE_NOT_FOUND;
 	}
 
-	return BOOT_ROM_SUCCESS;
+	/* Wait for previous fetch to complete */
+	boot_rom_boot_flow_wait_for_dma_finish();
+
+	return boot_rom_boot_flow_load_recovery_images(recovery_manifest, rec_img_count,
+							img_sts_base);
 }
 
 void execute_recovery(void)
